lib/Models/GnafApi.js

```diff
@@ -0,0 +1,311 @@
+ *      everywhere is searched.
+ * @param {number} [maxResults] The maximum number of results to return - defaults to 10.
+ * @returns {Promise} Promise that resolves when the search is done.
+ */
+GnafApi.prototype.geoCode = function(searchTerm, rectangle, maxResults) {
+    maxResults = maxResults || DEFAULT_MAX_RESULTS;
+
+    var requestData = buildRequestData(searchTerm, maxResults);
+
+    if (defined(rectangle)) {
+        addBoundingBox(requestData, rectangle);
+    }
+
+    return loadWithXhr({
+        url: this.url,
+        headers: {
+            Accept: 'application/json'
+        },
+        preferText: true,
+        method: 'POST',
+        data: JSON.stringify(requestData)
+    }).then(function(raw) {
+        var data = JSON.parse(raw);
+        var hits = data.hits.hits.map(convertElasticSearchHit.bind(undefined, defined(rectangle)));
+        // If we got anything good return it, otherwise search outside the location.
+        if (!defined(rectangle) || hits.length === maxResults) {
+            return hits;
+        } else {
+            var hitIds = hits.reduce(function(soFar, hit) {
+                soFar[hit.id] = true;
+                return soFar;
+            }, {});
+
+            return this.geoCode(searchTerm, undefined, maxResults).then(function(nonLocationHits) {
+                var filteredNonLocationHits = nonLocationHits.filter(function(hit) {
+                    return !hitIds[hit.id];
+                });
+
+                return hits.concat(filteredNonLocationHits).slice(0, maxResults);
+            });
+        }
+    }.bind(this));
+};
+
+/**
+ * Bulk geocodes in batches with a list of plain-text address. Returns results along with their scores.
+ *
+ * If a bounding box is provided, this is used and no results will be returned from outside the bounding box.
+ *
+ * @param {Array} searchTermList A list of strings representing the addresses to geocode
+ * @param {Rectangle} [rectangle] An optional rectangle describing a box to search within. If this is not provided then
+ *      everywhere is searched.
+ * @param {number} [batchSize=DEFAULT_BATCH_SIZE] Maximum number of addresses in each request.
+ * @param {number} [maxSimultaneousRequests=DEFAULT_MAX_SIMULTANEOUS_REQUESTS] Maximum number of requests submitted at one time.
+ * @returns {Promise} Promise that resolves when the search is done.
+ */
+GnafApi.prototype.bulkGeoCode = function(searchTermList, rectangle, batchSize, maxSimultaneousRequests) {
+    batchSize = batchSize || DEFAULT_BATCH_SIZE;
+    maxSimultaneousRequests = maxSimultaneousRequests || DEFAULT_MAX_SIMULTANEOUS_REQUESTS;
+    var searchTermBatches = splitIntoBatches(searchTermList, batchSize);
+    var searchTermBatchesRequests = splitIntoBatches(searchTermBatches, maxSimultaneousRequests);
+
+    var i = 0;
+    var mergedResults = [];
+    return this._runThrottledRequests(i, mergedResults, searchTermBatchesRequests, rectangle);
```

Comment:
Is this done recursively so you can time the batches properly? It's a good solution but it might be worth looking into having `runThrottledRequests` return the promise for the batch it's passed, then doing something like:

```
searchTermBatchesRequests
    .map(requests => this._runThrottledRequests(requests, rectangle))
    .reduce((accumulatedPromise, batchPromise) => accumulatedPromise.then(batchPromise), when())
    .then(resultsOfBatches => resultsOfBatches.reduce((acc, batchResult) => acc.concat(batchResult), []))
```

I'm happy with merging it like this too though.